DSA tag protocol support

yangbolu1991 opened the pull request:
This PR is to support DSA tag protocol in DSA core driver. Also added NXP NETC tag protocol driver, and enabled on NXP i.MX943 as an example to verify.

There is a hal dependency: https://github.com/zephyrproject-rtos/hal_nxp/pull/575
Thanks a lot.

Attached dsa sample log.
```
[00:00:00.056,000] <inf> phy_mii: PHY (2) ID 20005CA2
[00:00:00.107,000] <inf> phy_mii: PHY (3) ID 20005CA2
[00:00:00.144,000] <inf> dsa_netc: DSA user port 0 Link down
[00:00:00.145,000] <inf> dsa_netc: DSA user port 1 Link down
[00:00:00.145,000] <inf> nxp_imx_eth_psi: ENETC3 MAC: 00:04:9f:00:11:00
*** Booting Zephyr OS build v4.2.0-rc2-120-gb0fc94da3d01 ***
[00:00:00.155,000] <inf> net_dsa_sample: [0] User interface 1 found.
[00:00:00.155,000] <inf> net_dsa_sample: [1] User interface 2 found.
[00:00:00.155,000] <inf> net_dsa_sample: DSA ports init - OK
[00:00:02.214,000] <inf> phy_mii: PHY (3) Link speed 100 Mb, full duplex
[00:00:02.214,000] <inf> dsa_netc: DSA user port 1 Link up
[00:00:02.214,000] <inf> net_dsa_sample: Port 2 is up
uart:~$
uart:~$
[00:00:10.174,000] <inf> net_dhcpv4: Received: 10.193.20.37
uart:~$
uart:~$
uart:~$
uart:~$ net ping 10.193.20.164
PING 10.193.20.164
28 bytes from 10.193.20.164 to 10.193.20.37: icmp_seq=1 ttl=64 time=0 ms
28 bytes from 10.193.20.164 to 10.193.20.37: icmp_seq=2 ttl=64 time=0 ms
28 bytes from 10.193.20.164 to 10.193.20.37: icmp_seq=3 ttl=64 time=0 ms
[00:00:23.722,000] <inf> phy_mii: PHY (3) is down
[00:00:23.722,000] <inf> dsa_netc: DSA user port 1 Link down
[00:00:28.163,000] <inf> phy_mii: PHY (2) Link speed 100 Mb, full duplex
[00:00:28.163,000] <inf> dsa_netc: DSA user port 0 Link up
[00:00:28.163,000] <inf> net_dsa_sample: Port 1 is up
uart:~$
uart:~$
uart:~$
[00:00:34.122,000] <inf> net_dhcpv4: Received: 10.193.20.163
uart:~$ net ping 10.193.20.164
PING 10.193.20.164
28 bytes from 10.193.20.164 to 10.193.20.163: icmp_seq=1 ttl=64 time=1 ms
28 bytes from 10.193.20.164 to 10.193.20.163: icmp_seq=2 ttl=64 time=0 ms
28 bytes from 10.193.20.164 to 10.193.20.163: icmp_seq=3 ttl=64 time=0 ms
uart:~$

```

yangbolu1991 committed:
net: dsa: move DSA core options to subsystem

Moved DSA core options to subsystem.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: move NXP NETC DSA driver to dsa directory

Moved NXP NETC DSA driver to dsa directory.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: dsa: add tag protocol definition and binding

Added tag protocol definition and binding. Also introduced
zephyr,dsa-port compatible for future tag protocol driver Kconfig
dependency checking.

Updated existed dts.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: dsa: add NXP NETC tag protocol driver

Added NXP NETC tag protocol driver.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: dsa: support tag protocol setup

Supported tag protocol setup.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: eth_nxp_imx_netc: allow to transmit DSA tag frame

Allowed to transmit DSA tag frame.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: dsa_nxp_imx_netc: add tag protocol support

Added tag protocol support.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: imx943_evk: add NETC switch support for M33

Added NETC switch support for M33.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
boards: nxp: imx943_evk: doc update for NETC switch support for M33

Added NETC switch support for M33 in doc.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: dsa: add imx943_evk_mimx94398_m33_ddr board support

Added imx943_evk_mimx94398_m33_ddr board support.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
west.yml: update hal_nxp to support netc tag driver

Updated hal_nxp to support netc tag driver.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to fix twister issues.
Thanks.

JiafeiPan commented on include/zephyr/net/dsa_core.h:
```diff
@@ -75,6 +75,12 @@ extern "C" {
 	};                                                                                         \
 	DT_INST_FOREACH_CHILD_STATUS_OKAY_VARGS(n, fn, n);
 
+#define DSA_TAG_PROTO_NONE	0
```
change it to be:
`#define DSA_TAG_PROTO_NOTAG	0` ?

yangbolu1991 replied:
Okay, thanks.

JiafeiPan commented on subsys/net/l2/ethernet/dsa/dsa_tag_netc.h:
```diff
@@ -0,0 +1,24 @@
+static inline struct net_pkt *dsa_tag_netc_xmit(struct net_if *iface, struct net_pkt *pkt)
+{
+	return pkt;
+}
```
if not defined CONFIG_DSA_TAG_PROTOCOL_NETC, it should be notag for netc switch, so no need to define these two functions in "else" branch for each tag driver, the common [tag drive](https://github.com/zephyrproject-rtos/zephyr/pull/93010/files#diff-7d67425b02144b583c1670b7083e08abad1e2fc902181d6d1b9dddcc4a274467R35) will handle notag for all DSA drivers.

yangbolu1991 replied:
Okay.
Let me change to use conditional compile to reduce resource as possible.
Thanks.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -75,6 +75,12 @@ extern "C" {
 	DT_INST_FOREACH_CHILD_STATUS_OKAY_VARGS(n, fn, n);
 
+#define DSA_TAG_PROTO_NONE	0
+#define DSA_TAG_PROTO_NETC	1
```
Okay, thanks.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -75,6 +75,12 @@ extern "C" {
+#define DSA_TAG_PROTO_NETC	1
+
+#define DSA_TAG_PROTO_BY_DT(node_id) \
+	(DT_ENUM_HAS_VALUE(node_id, tag_protocol, netc) ? DSA_TAG_PROTO_NETC : DSA_TAG_PROTO_NONE)
```
Okay, thanks.

yangbolu1991 commented on drivers/ethernet/dsa/dsa_nxp_imx_netc.c:
```diff
@@ -128,6 +128,7 @@ static struct dsa_api dsa_netc_api = {
 		.port_idx = DT_REG_ADDR(port),                                              \
 		.phy_dev = DEVICE_DT_GET_OR_NULL(DT_PHANDLE(port, phy_handle)),             \
 		.phy_mode = DT_PROP_OR(port, phy_connection_type, ""),                      \
+		.tag_proto = DSA_TAG_PROTO_BY_DT(port),                                     \
```
Okay, thanks.

yangbolu1991 commented on subsys/net/l2/ethernet/dsa/dsa_tag.h:
```diff
@@ -6,7 +6,10 @@
 #ifndef ZEPHYR_SUBSYS_DSA_TAG_PRIV_H_
 #define ZEPHYR_SUBSYS_DSA_TAG_PRIV_H_
 
+#include "dsa_tag_netc.h"
```
I think choosing conditional compiling, or choosing to use stub functions are two ways.
I am using stub functions to avoid a lot #if in header files, or c files.

Thanks.

yangbolu1991 replied:
Let me change to use conditional compile to reduce resource as possible.
Thanks.

yangbolu1991 commented on subsys/net/l2/ethernet/dsa/dsa_tag.c:
```diff
@@ -36,3 +37,20 @@ struct net_pkt *dsa_tag_xmit(struct net_if *iface, struct net_pkt *pkt)
+	case DSA_TAG_PROTO_NETC:
+		dsa_switch_ctx->dapi->recv = dsa_tag_netc_recv;
+		dsa_switch_ctx->dapi->xmit = dsa_tag_netc_xmit;
+		break;
```
No need to use `#if` as stub functions are used.
Will add no tag checking.
Thanks.

clamattia commented:
@yangbolu1991 Please if possible use the `Start review` feature instead of `Add single comment` to avoid spamming subscribers.

clamattia commented:
> > @yangbolu1991 Please if possible use the `Start review` feature instead of `Add single comment` to avoid spamming subscribers.
> 
> Get it. And sorry for the trouble...

No problem and thank you for the great work.

pdgendt commented on dts/bindings/dsa/dsa.yaml:
```diff
@@ -4,7 +4,9 @@
-  description: Properties of slave port
+  description: Properties of port
+
+  compatible: "dsa-port"
```
Won't this throw warnings of an unknown/missing vendor?

```suggestion
  compatible: "zephyr,dsa-port"
```

pdgendt commented on include/zephyr/dt-bindings/ethernet/dsa_tag_proto.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_DSA_TAG_PROTO_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_DSA_TAG_PROTO_H_
+
+#define DSA_TAG_PROTO_NOTAG	0
+#define DSA_TAG_PROTO_NETC	1
```
Some comments describing the defines would be helpful.

yangbolu1991 replied:
Thanks. Will add that.

pdgendt commented on subsys/net/l2/ethernet/dsa/dsa_tag_netc.c:
```diff
@@ -0,0 +1,96 @@
+
+	/* allocate new net_buf */
+	buf = net_buf_alloc_len(net_buf_pool_get(pkt->buffer->pool_id), buf_len, K_NO_WAIT);
+	if (!buf) {
```
```suggestion
	if (buf == NULL) {
```

yangbolu1991 replied:
Thanks, will fix.

yangbolu1991 commented:
Updated to v3. Changes include,
- Code improvement per @pdgendt suggestion.

Thanks.

yangbolu1991 committed:
dts: arm: nxp_imx943_m33: add NETC switch node

Added NETC switch node.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Rebased to address conflict.
Thanks.

JiafeiPan commented on subsys/net/l2/ethernet/dsa/dsa_tag_netc.c:
```diff
@@ -0,0 +1,96 @@
+	/* copy the rest data */
+	memcpy((void *)((uintptr_t)buf->data + sizeof(*header)),
+	       (void *)((uintptr_t)pkt->frags->data + (2 * NET_ETH_ADDR_LEN)),
+	       pkt->frags->len - (2 * NET_ETH_ADDR_LEN));
```
Is this to copy the whole frame data to new buffer? It will impact the performance, in future we could reserve some header room in the buffer, so that we could only move some frame header fields before tag field into buffer header room and insert the tag, no need to move or copy frame data which maybe with huge date length.
No blocking for this PR, just some performance optimization suggestions.

yangbolu1991 replied:
Right. I think there will be some work to do in NET driver for that.
But let me update code to just insert a frag of tag header and avoid data copying.

Thanks.

yangbolu1991 commented:
Updated to v4. Changes include,
- Addressed conflict.
- Avoided data copying during netc tag header inserting.

Thanks.

yangbolu1991 committed:
test: ram_context_for_isr: exclude imx943 m33

i.MX943 M33 supports multi-level interrupt with NUM_IRQS 405.
Current ram_context_for_isr test does not cover multi-level interrupt.
So exclude i.MX943 M33.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
It seems code base introduced a new test case which didn't apply to i.MX943 M33, causing twister issues.
So, added one more patch to exclude i.MX943 M33 to fix twister issues.

Thanks.

yangbolu1991 commented:
No changes. Just re-tigger CI as last CI result showed some scripts problem.
Thanks.

maass-hamburg commented on subsys/net/l2/ethernet/dsa/dsa_tag_netc.c:
```diff
@@ -0,0 +1,89 @@
+/*
```
@rlubos @jukkar what do you think, as you both are the maintainers of that subsystem?

jukkar commented on subsys/net/l2/ethernet/dsa/dsa_tag.c:
```diff
@@ -36,3 +40,24 @@ struct net_pkt *dsa_tag_xmit(struct net_if *iface, struct net_pkt *pkt)
+	struct dsa_switch_context *dsa_switch_ctx = dev_cpu->data;
+
+	switch (cfg->tag_proto) {
+#ifdef CONFIG_DSA_TAG_PROTOCOL_NETC
```
We could enhance and make this more generic in the future by having a macro that setups the tag for a vendor and places the information to a linker section. Then the generic setup function would just traverse the dsa tag section and config things accordingly.
This way we would not need to edit this file at all if other vendor protocols are added and this would also support out of tree vendor dsa tags. Anyway, this is something that can be implemented in the future.

yangbolu1991 replied:
Thanks @jukkar for the good suggestion.
That's a good way to manage tag protocols. I will try to convert to this way in the future.
